pipelines/build/common/build_base_file.groovy

```diff
@@ -263,27 +265,58 @@ class Builder implements Serializable {
     This function parses and applies this to the individual build config.
     */
     Map<String, ?> getDynamicParams(Map<String, ?> configuration, String variant) {
-        List<String> testLists = DEFAULTS_JSON["testDetails"]["defaultDynamicParas"]["testLists"]
-        String numMachines = DEFAULTS_JSON["testDetails"]["defaultDynamicParas"]["numMachines"]
+        List<String> testLists = []
+        List<String> numMachines = []
+
+        def testDynamicMap
+
         if (configuration.containsKey("testDynamic")) {
-            if (configuration.get("testDynamic")) {
-                if(configuration.get("testDynamic").containsKey(variant)) {
-                    testLists = configuration.get("testDynamic").get(variant).get("testLists")
-                    numMachines = configuration.get("testDynamic").get(variant).get("numMachines")
-                } else {
-                    testLists = configuration.get("testDynamic").get("testLists")
-                    numMachines = configuration.get("testDynamic").get("numMachines")
+            // fetch from buildConfigurations for target
+
+            // testDynamic could be map, list or boolean
+            if (configuration.containsKey("testDynamic") && configuration.get("testDynamic")) {
+                // fetch variant options
+                testDynamicMap = configuration.get("testDynamic").get(variant)
+            } else {
+                // fetch generic options
+                testDynamicMap = configuration.get("testDynamic")
+            }
+        } else if (DEFAULTS_JSON["testDetails"]["defaultDynamicParas"]) {
+            // fetch default options
+            testDynamicMap = DEFAULTS_JSON["testDetails"]["defaultDynamicParas"]
+        }
+
+        if (testDynamicMap.containsKey("testLists")) {
+            testLists.addAll(testDynamicMap.get("testLists"))
+        }
+
+        if (testDynamicMap.containsKey("numMachines")) {
+            // populate the list of number of machines per tests
+            if (List.class.isInstance(testDynamicMap.get("numMachines"))) {
+                // the size of the numMachines List should match the testLists size
+                // otherwize throw an error
+                // e.g. 
+                // testLists    = ['extended.openjdk', 'extended.jck', 'special.jck']
+                // numMachines  = ['3',                '2',            '5']
+
+                numMachines.addAll(testDynamicMap.get("numMachines"))
+            } else {
+                if (!testLists.isEmpty()) {
+                    // work-around for numMachines as a String
+                    // populate the List<String> number of machines with duplicates
+                    numMachines.addAll(Collections.nCopies(testLists.size(), "${testDynamicMap.get('numMachines')}"))
                 }
-           } else {
-                testLists = []
-                numMachines = ""
             }
         }
-        
+
+        if (numMachines.size() < testLists.size()) {
```

Comment:
Shouldn't this be `!=`?

Reply:
This check is sufficient to avoid a numMachines.getAt(index) != null

Reply:
So, what are the valid cases?
numMachines.size = 1 or numMachines.size = testLists.size?
Shouldn't this check be validating those scenarios? Or am I misunderstanding?

Reply:
The only case we want to fail is numMachines.size < testLists.size.
numMachines.size >= testLists.size is a valid case